src/manual/devel/best-practices/microservices.md

```diff
@@ -0,0 +1,307 @@
+This shared responsibility is easier for everyone if the implementations are similar, avoiding surprises when moving from one service to another.
+This document aims to collect the practices and standards we've agreed on.
+
+Most services do not live up to these standards -- TaskCluster is a work in progress -- but pull requests to bring services up-to-date with one or more points are always appreciated.
+Many of these aren't hard-and-fast rules, either.
+If there's good reason to do otherwise, do so!
+
+## Package Mechanics
+
+### Repository
+
+Services should be in a Github repository in the `taskcluster` organization, with the repo name having the prefix `taskcluster-`.
+
+### Source Layout
+
+Include all source in the `src/` directory, and all tests in `test/`.
+The babel-compile configuration will compile those to `lib/` and `.test/`, respectively.
+Within the `src` directory, the main script should be named `main.js`.
+
+### Compiling
+
+Use [babel-compile](https://github.com/taskcluster/babel-compile) to compile your application.
+Its README describes how to set it up generally, but for a TaskCluster service, you will need to install `babel-compile` and `babel-preset-taskcluster`.
+Include the following in your `package.json`:
+
+```js
+  "scripts": {
+    "compile": "babel-compile -p taskcluster src:lib test:.test",
+    "pretest": "npm run compile",
+    "install": "npm run compile",
+  }
+```
+
+### Dependencies
+
+Try to keep up-to-date with the latest versions of all TaskCluster libraries.
+In general, the implications of updating these libraries should be clear, and the authors are easy to find when things go badly.
+
+Other dependencies should also be kept up-to-date as much as possible.
+Tools like [Greenkeeper](https://greenkeeper.io/) can be very useful for this purpose.
+
+### Shrinkwrap
+
+Dependencies for deployed services are handled with [`npm shrinkwrap`](https://docs.npmjs.com/cli/shrinkwrap).
+This is to ensure that we are deploying a consistent tree of package versions, and that dependency upgrades are controlled and not automatic or surprising.
+
+When creating a new service, run `npm shrinkwrap` to generate an `npm-shrinkwrap.json` based on the packages you have installed in your development environment, then check that file into the Git repository.
+When beginning development on an existing service, the usual `npm install` will install the specific versions named in the `npm-shrinkwrap.json`.
+
+*Important*: Do not use `devDependencies`, as npm versions before 3.0 do not deal with them correctly and will cause failures in deployment that do not appear in local runs or Travis.
+Instead, list all dependencies in `dependencies`.
+
+When changing a service's dependencies, the easiest approach is to use `npm install --save` to install the package and update `package.json` at the same time.
+Once this is complete, run `npm shrinkwrap` and add the modified `npm-shrinkwrap.json` to your commit.
```

Comment:
yes, it will.  We should also mention that removing packages should be done with NPM as well, since it will also update shrinkwrap correctly to remove them.  And that package upgrades should be done like `npm install dep@...` instead of just editing package.json so that shrinkwrap is updated.